### PR TITLE
Fixes row like

### PR DESCRIPTION
**Pull request by chambart**
It does not fix the row like meet itself.
The main problems fixed are the confusion around Product that does represent the bottom.
And the handling of environment extension around disjunctions.

When meeting (A U B) and (C U D) the meet results in
M1, env_ext1 = meet A C
M2, env_ext2 = meet B D

The value is M1 U M2, but the env_extension was env_ext1 meet env_ext2. But this should be a join instead.
Given the way the meet currently works with the extension, this is very imprecise. For instance
meet (A a U B a) and (A 1 U B unknown) used to return (A a U B a) { a = 1 } which is wrong.
now it returns (A a U B a) { a = unknown } which lost the information about the A case it could have also been (A 1 U B a) { a = unknown }

### CODE REVIEW

File: middle_end/flambda2.0/types/structures/closures_entry.rec.ml [hunk 1]
```diff
@@ -35,14 +38,14 @@ let create ~function_decls ~closure_types ~closure_var_types =
...
 let create_bottom () =
   { function_decls = Closure_id.Map.empty;
-    closure_types = PC.create_bottom ();
-    closure_var_types = PV.create_bottom ();
+    closure_types = PC.create_top Flambda_kind.value;
+    closure_var_types = PV.create_top Flambda_kind.value;
   }
 
-let is_bottom { function_decls; closure_types; closure_var_types; } =
+let is_bottom { function_decls; closure_types = _; closure_var_types = _; } =
```
Comment (chambart) [inline middle_end/flambda2.0/types/structures/closures_entry.rec.ml]: This is wrong. I think that this can't really represents bottom.

**Comment by lpw25** [comment]
Interestingly this issue -- which I like to think of as a principality issue -- wasn't present in the earlier prototype that allowed equations underneath constructors because it could represent the solution as (A { a = 1 } a U B a).

As a side-note, I suspect we would rather have (A 1 U B a) { a = unknown } than
(A a U B a) { a = unknown } since more optimizations care about the concrete representation than the relationships between things, but I don't know how easy it would be to choose those solutions consistently.

**Comment by chambart** [comment]
I have tried adding env_extension in row_like cases. It helps with that problem and stores something similar to `(A { a = 1 } a U B a)`. It seems to do the right thing but I don't know yet how expensive it is to maintain.

I wont push it in this PR as it needs some fixes on join of env_extension.

**Comment by chambart** [comment]
With the last patches this seems to work quite well. I could pass the testsuite with ocamlopt.opt

**Comment by lpw25** [comment]
> I have tried adding env_extension in row_like cases. It helps with that problem and stores something similar to (A { a = 1 } a U B a). It seems to do the right thing but I don't know yet how expensive it is to maintain.

That's what our old prototype did, but we deliberately chose not to include that in this version because we're worried about the cost. I suspect it would be better to leave it out for now and accept the loss in precision. Once we've got this version working well then we can look at re-introducing these extensions and see if they can be done without things blowing up.

**Comment by chambart** [comment]
Rebased on @Gbury 's branch fixing pack

**Comment by chambart** [comment]
This has been merged in flambda2.0-speed which replaced -stable. Hence was hand merged.

**Pull request closed without merging**